compiler/lib/inline.ml

```diff
@@ -21,129 +21,457 @@
+
+let body_size ~context info = cache ~info info.body_size (size ~recurse:false ~context)
+
+let full_size ~context info = cache ~info info.full_size (size ~recurse:true ~context)
+
+let closure_count_uncached ~context =
+  sum ~context (fun { body; _ } ->
+      List.fold_left
+        ~f:(fun n i ->
+          match i with
+          | Let (_, Closure _) -> n + 1
+          | _ -> n)
+        ~init:0
+        body)
+
+let closure_count ~context info =
+  cache ~info info.closure_count (closure_count_uncached ~context)
+
+let count_init_code ~context info =
+  cache
+    ~info
+    info.init_code
+    (sum ~context
+    @@ fun { body; _ } ->
+    List.fold_left
+      ~f:(fun n i ->
+        match i with
+        | Let (_, (Closure _ | Field _ | Constant _ | Block _)) -> n + 1
+        | Let (_, (Apply _ | Prim _ | Special _))
+        | Assign _ | Set_field _ | Offset_ref _ | Array_set _ | Event _ -> n)
+      ~init:0
+      body)
+
+let returns_a_block ~context info =
+  cache ~info info.returns_a_block (fun pc ->
+      let blocks = context.p.blocks in
+      Code.traverse
+        { fold = fold_children }
+        (fun pc acc ->
+          acc
+          &&
+          let block = Addr.Map.find pc blocks in
+          match block.branch with
+          | Return x -> (
+              match Code.last_instr block.body with
+              | Some (Let (x', Block _)) -> Var.equal x x'
+              | _ -> false)
+          | Raise _ | Stop | Branch _ | Cond _ | Switch _ | Pushtrap _ | Poptrap _ -> true)
+        pc
+        blocks
+        true)
+
+let interesting_parameters ~context info =
+  let params = info.params in
+  cache ~info info.interesting_params (fun pc ->
+      let params = List.filter ~f:(fun x -> context.live_vars.(Var.idx x) = 1) params in
+      if List.is_empty params
+      then []
+      else
+        let blocks = context.p.blocks in
+        Code.traverse
+          { fold = fold_children }
+          (fun pc lst ->
+            let block = Addr.Map.find pc blocks in
+            List.fold_left
+              ~f:(fun lst i ->
+                match i with
+                | Let (_, Apply { f; args; _ }) when List.mem f ~set:params ->
+                    (f, List.length args) :: lst
+                | _ -> lst)
+              ~init:lst
+              block.body)
+          pc
+          blocks
+          [])
+
+(*
+   We are very aggressive at optimizing functor-like code, even if
+   this might duplicate quite a lot of code, since this is likely to
+   allow other optimizations: direct function calls, more precise dead
+   code elimination, ...
+*)
+let functor_like ~context info =
+  (context.aggressive || body_size ~context info <= 15)
+  && (not info.recursive)
+  && (not (contains_loop ~context info))
+  && returns_a_block ~context info
+  && count_init_code ~context info * 2 > body_size ~context info
+  && full_size ~context info - body_size ~context info <= 20 * closure_count ~context info
```

Comment:
maybe add a small comment explaining the two heuristics above , `init_code * 2 > body_size` and `full_size <= 20 * closure_count`

Reply:
@vouillon, just making sure you saw this comment